client/shared/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -289,16 +291,40 @@ object Client {
       go(source).stream
     }
 
-    def runOrDispose(req: Request[F]): F[Resource[F, Response[F]]] =
-      Ref[F].of(false).map { disposed =>
-        val req0 = addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
-        Resource
-          .eval(app(req0))
-          .onFinalize(disposed.set(true))
-          .map(_.pipeBodyThrough(until(disposed)))
-      }
+    def run(
+        message: Message[F],
+        refOp: Option[Ref[F, Boolean]] = None,
+    ): Resource[F, Response[F]] = message match {
+      case req: Request[F] =>
+        Resource.eval(Ref[F].of(false)).flatMap { disposed =>
+          val reqAugmented =
+            addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
+          Resource
+            .eval(app(reqAugmented))
+            .onFinalize(disposed.set(true))
+            .flatMap(run(_, Option(disposed)))
+        }
+      case resp: Response[F] =>
+        for {
+          disposed <- Resource.eval(refOp.fold(Ref[F].of(false))(F.pure))
+          channel <- Resource.make(Channel.unbounded[F, Byte])(_.close.void)
+          r = resp.withBodyStream(
+            channel.stream
+              .through(until(disposed)(_))
+              .onFinalize(channel.stream.compile.drain.guarantee(disposed.set(true)))
+              .concurrently(channel.sendAll(resp.body))
```

Comment:
I'm not entirely sure if this is right, the issue with `concurrently` is that it will be interrupted as soon as the the primary stream is interrupted. And actually we don't want the `resp.body` stream to _ever_ be interrupted, that is the whole point of directing it into the `Channel`.

Reply:
Please correct me if I'm wrong, but if the logic replaces the response stream with a new stream `resp.withBodyStream`, then the client should not be able to interrupt the original `resp.body`. 

I understood your comment about using `Channel` that the client can forward bytes from `resp.body` to `Channel` and use `channel.stream` as a body stream. And if the user interrupts reading the stream from the step where client stopped due to the `Channel` implementation the logic can drain the rest of bytes from the point where the user stopped. 

Did I understand you correctly?

Reply:
Let me "zoom" on the problem:
```scala
channel.stream.concurrently(channel.sendAll(resp.body))
```
In this case, if the client stops consuming the stream, then it will also interrupt the `concurrently` "background" stream.

> The resulting stream terminates upon termination of this stream. The background stream will be interrupted at that point.

https://github.com/typelevel/fs2/blob/24ea712f61d3276cc5798ffa00396a7fa9378d30/core/shared/src/main/scala/fs2/Stream.scala#L519-L520

Since the "background" stream is based on `resp.body`, that means that `resp.body` will be interrupted.

Let me know if that makes sense!

Reply:
Thanks for the clarification. I've added a unit test to cover this case.